### ISSUE TITLE
RangeError: attempting to construct out-of-bounds TypedArray on ArrayBuffer

### ISSUE THREAD

**Issue by JanisE** [photopea/UPNG.js#25]
Hello! I'm trying to use UPNG.js to optimise the PNG image returned by Canvas, like this:

```javascript
function OpenOptimisedImage(canvas) {
    canvas.toBlob((imageBlob) => {
        const fileReader = new FileReader();

        fileReader.onloadend = (canvasConvertedToArrayBuffer) => {
            var img  = window.UPNG.decode(
                // 0 = losslessly.
                window.UPNG.encode([canvasConvertedToArrayBuffer.target.result], canvas.width, canvas.height, 0)
            );
            var rgba = window.UPNG.toRGBA8(img)[0];
            var blob = new Blob([rgba], {type:'application/octet-binary'});

            var reader = new FileReader();

            reader.onloadend = (optimisedBlobConvertedToBase64) => {
                window.open(optimisedBlobConvertedToBase64.target.result);
            };
            reader.readAsDataURL(blob);
        };

        fileReader.readAsArrayBuffer(imageBlob);
    });
}
```

I get an error "RangeError: attempting to construct out-of-bounds TypedArray on ArrayBuffer" on line
```
UPNG.encode.compress [..]
var img32 = new Uint32Array(bufs[j]), ilen = img32.length;
```
if I use the 2.1 version from `npm install` or https://cdnjs.cloudflare.com/ajax/libs/upng-js/2.1.0/UPNG.js, or on line
```
UPNG.encode.framize [..]
var cimg = new Uint8Array(bufs[j]), cimg32 = new Uint32Array(cimg.buffer);
```
if I'm using the latest `UPNG.js` as it is in the repository now.

It seems that `Uint8Array` can be created, but not `Uint32Array`.

Only once I got past the first `UPNG.encode` call, and I'm not sure why the error didn't happen then.

What is the problem?

**Comment by JanisE** [comment]
Yes, it is. It outputs:
`ArrayBuffer { byteLength: 114038 }`

BTW, my first try after your comment I got a second case when the error didn't appear. The result was not what I wanted, but that will be another problem. On the second try I got the error again. (Probably depends on the specific canvas image, which is dynamic and different from one case to another.)

**Comment by photopea** [comment]
The length of your arraybuffer 114038 is not divisible by 4 . The arrayBuffer should contain RGBA components for each pixel. I.e. there must be 4 bytes per pixel - your raw data should be a multiple of 4.

**Comment by photopea** [comment]
You can crate an image manually to see if it works

UPNG.encode([(new Uint8Array([0,0,0,255, 255,0,0,255])).buffer], 2, 1, 0) - this should make a two-pixel image with a black and a red pixel

**Comment by JanisE** [comment]
Thank you for the pointer!

Not knowing all the formats being at play here, it kept me experimenting and trying things out for quite some time just to optimise a Canvas image.

I've made a pull request #26 of the result, which I hope will be useful to others trying similar things.

**Comment by photopea** [comment]
I added a sample code into Readme about how to read from canvas.

**Comment by JanisE** [comment]
Thanks!

Maybe you could also add this part, because it may not be clear how the resulting `png` can be used in its given format:

```javascript
new Promise((resolve) => {
    var reader = new FileReader();
    reader.onloadend = (pngConvertedToBase64) => {
        resolve(pngConvertedToBase64.target.result);
    };
    reader.readAsDataURL(new Blob([png], {type: 'image/png'}));
}).then((pngB64) => window.open(pngB64));
```

**Comment by photopea** [comment]
The PNG file could be sent somewhere through the network, or attached to some other file (like ZIP archive). These scenarios are out of scope of this library.
